proposals/0072-feature-gate-threshold-automation.md

```diff
@@ -0,0 +1,248 @@
+---
+simd: '0072'
+title: Feature Gate Threshold Automation
+authors:
+  - Tyera Eulberg
+  - Joe Caulfield
+category: Standard
+type: Core
+status: Idea
+created: 2024-01-25
+feature: (fill in with feature tracking issues once accepted)
+---
+
+## Summary
+
+This SIMD outlines a proposal for automating the feature activation process
+based on a stake-weighted support threshold, rather than manual human action.
+
+With this new process, contributors no longer have to assess stake support for a
+feature before activation. Instead, the assessment is done by the runtime.
+
+## Motivation
+
+Feature gates wrap new cluster functionality, and typically change the rules of
+consensus. As such, a feature gate needs to be supported by a strong majority
+of cluster stake when it is activated, or else it risks partitioning the
+network. The current feature-gate process involves two steps:
+
+1. An individual key-holder stages a feature gate for activation
+2. The runtime automatically activates the feature on the next epoch boundary
+
+The key-holder is the one who *manually* (with the help of the solana-cli)
+assesses the amount of stake that recognizes the feature and decides whether
+it is safe to activate. This is obviously brittle and subject to human error.
+
+If instead the runtime was to assess this stake support for activating a
+feature, this would eliminate the key-holder's responsibility to asses stake
+support, reducing the risk of human error.
+
+In a world where multiple clients will aim to push features and seek to agree on
+their activation, a more automated and secure process will be extremely
+beneficial.
+
+## New Terminology
+
+- **Staging authority**: The authority who must sign to stage a given feature
+  for activation, or unstage that feature.
+- **Signal account (PDA)**: A PDA under the Feature Gate program used to
+  prevent double-signaling.
+
+## Detailed Design
+
+A new feature activation process, comprised of the following steps:
+
+1. Contributor creates a feature account and configures the feature's staging
+   authority.
+2. In some epoch `N-1`, the staging authority stages the feature for activation.
+3. During the next epoch `N`, validators signal which of the staged features
+   they support.
+4. At the end of epoch `N`, the runtime activates the features with enough stake
+   support.
+
+### New Feature Account State
+
+A new version of feature account state will track feature status, staging
+authority, and stake support. Since the current feature state is a Rust
+`Option<u64>`, we will use a `2` for the first byte to denote feature state v2.
+
+```rust
+enum FeatureV2Status {
+    Inactive {
+        staging_authority: Pubkey,
+    },
+    Staged { 
+        staging_authority: Pubkey,
+        stake_support: u64,
+    },
```

Comment:
The staged state should also have the current epoch number so that at epoch boundaries we know that support was signaled in the previous epoch and not during an earlier epoch. This is important because the total stake can change epoch to epoch and affect whether at epoch boundaries a certain amount of stake is sufficient for activation.